File: packages/odoo-client/src/odoo_client/widget_catalog.py
"""Curated Odoo field widgets per ttype (compendium §4 — not the full registry)."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class WidgetOption:
    id: str
    label: str
    hint: str = ""


WIDGETS_BY_TTYPE: dict[str, tuple[WidgetOption, ...]] = {
    "char": (
        WidgetOption("email", "Email", "widget=email"),
        WidgetOption("phone", "Phone", "widget=phone"),
        WidgetOption("url", "URL", "widget=url"),
        WidgetOption("barcode", "Barcode", "widget=barcode"),
    ),
    "integer": (
        WidgetOption("priority", "Priority", "widget=priority"),
    ),
    "float": (
        WidgetOption("float_time", "Float time", "widget=float_time"),
        WidgetOption("progressbar", "Progress bar", "widget=progressbar"),
        WidgetOption("percentage", "Percentage", "widget=percentage"),
    ),
    "selection": (
        WidgetOption("radio", "Radio", "widget=radio"),
        WidgetOption("priority", "Priority", "widget=priority"),
        WidgetOption("selection_badge", "Badge", "widget=selection_badge"),
    ),
    "many2many": (
        WidgetOption("many2many_tags", "Tags", "widget=many2many_tags"),
        WidgetOption("many2many_checkboxes", "Checkboxes", "widget=many2many_checkboxes"),
    ),
    "many2one": (
        WidgetOption("many2one_avatar", "Avatar", "widget=many2one_avatar"),
        WidgetOption("many2one_avatar_user", "User avatar", "widget=many2one_avatar_user"),
    ),
    "binary": (
        WidgetOption("image", "Image", "widget=image"),
        WidgetOption("pdf_viewer", "PDF viewer", "widget=pdf_viewer"),
        WidgetOption("signature", "Signature", "widget=signature"),
    ),
    "html": (
        WidgetOption("html", "HTML", "widget=html"),
    ),
}


def widgets_for_ttype(ttype: str) -> tuple[WidgetOption, ...]:
    return WIDGETS_BY_TTYPE.get(ttype, ())


def validate_widget_for_ttype(ttype: str, widget: str | None) -> str | None:
    if not widget:
        return None
    from odoo_client.niche_widget_catalog import NICHE_WIDGETS

    allowed = {w.id for w in widgets_for_ttype(ttype)}
    for nw in NICHE_WIDGETS:
        if ttype in nw.recommended_ttypes:
            allowed.add(nw.id)
    if widget not in allowed:
        raise ValueError(f"Widget {widget!r} is not in the curated set for ttype {ttype!r}")
    return widget

File: apps/web/src/lib/widgetCatalog.ts
/** Curated Odoo field widgets (mirrors packages/odoo-client widget_catalog.py). */

export type WidgetOption = {
  id: string;
  label: string;
  hint?: string;
};

export const IMAGE_SIZE_PRESETS: { label: string; options: string }[] = [
  { label: "90 × 90", options: '{"size": [90, 90]}' },
  { label: "128 × 128", options: '{"size": [128, 128]}' },
  { label: "256 × 256", options: '{"size": [256, 256]}' },
];

export const WIDGET_OPTION_PRESETS: Record<string, { label: string; options: string }[]> = {
  image: IMAGE_SIZE_PRESETS,
  many2many_tags: [
    { label: "No create", options: '{"no_create": true}' },
    { label: "Color from color field", options: '{"color_field": "color"}' },
  ],
  radio: [{ label: "Horizontal", options: '{"horizontal": true}' }],
  phone: [{ label: "Enable SMS", options: '{"enable_sms": true}' }],
  progressbar: [{ label: "Current / max", options: '{"max_value": "x_max"}' }],
};

export const CSS_CLASS_PRESETS: { value: string; label: string }[] = [
  { value: "oe_inline", label: "Inline" },
  { value: "oe_avatar", label: "Avatar" },
  { value: "oe_highlight", label: "Highlight" },
  { value: "o_address_type", label: "Address type" },
];

export const GROUP_XML_PRESETS: { xmlId: string; label: string }[] = [
  { xmlId: "base.group_user", label: "Internal user" },
  { xmlId: "base.group_system", label: "Settings" },
  { xmlId: "base.group_erp_manager", label: "Access rights" },
  { xmlId: "base.group_portal", label: "Portal" },
];

export const PLACEHOLDER_TTYPES = new Set([
  "char",
  "text",
  "html",
  "many2one",
  "integer",
  "float",
  "monetary",
  "date",
  "datetime",
]);

export type GroupVisibilityToken = { xmlId: string; forbid: boolean };

export function parseGroupsAttr(raw: string | undefined): GroupVisibilityToken[] {
  if (!raw?.trim()) return [];
  return raw
    .split(",")
    .map((part) => part.trim())
    .filter(Boolean)
    .map((tok) => {
      const forbid = tok.startsWith("!");
      return { xmlId: forbid ? tok.slice(1).trim() : tok, forbid };
    })
    .filter((t) => t.xmlId);
}

export function serializeGroupsAttr(tokens: GroupVisibilityToken[]): string | undefined {
  const cleaned = tokens
    .map((t) => ({ xmlId: t.xmlId.trim(), forbid: t.forbid }))
    .filter((t) => t.xmlId);
  if (!cleaned.length) return undefined;
  return cleaned.map((t) => (t.forbid ? `!${t.xmlId}` : t.xmlId)).join(",");
}

const FALLBACK: Record<string, WidgetOption[]> = {
  char: [
    { id: "email", label: "Email" },
    { id: "phone", label: "Phone" },
    { id: "url", label: "URL" },
    { id: "barcode", label: "Barcode" },
  ],
  integer: [{ id: "priority", label: "Priority" }],
  float: [
    { id: "float_time", label: "Float time" },
    { id: "progressbar", label: "Progress bar" },
    { id: "percentage", label: "Percentage" },
  ],
  selection: [
    { id: "radio", label: "Radio" },
    { id: "priority", label: "Priority" },
    { id: "selection_badge", label: "Badge" },
  ],
  many2many: [
    { id: "many2many_tags", label: "Tags" },
    { id: "many2many_checkboxes", label: "Checkboxes" },
  ],
  many2one: [
    { id: "many2one_avatar", label: "Avatar" },
    { id: "many2one_avatar_user", label: "User avatar" },
  ],
  binary: [
    { id: "image", label: "Image" },
    { id: "pdf_viewer", label: "PDF viewer" },
    { id: "signature", label: "Signature" },
  ],
  html: [{ id: "html", label: "HTML" }],
};

export function fallbackWidgetsForTtype(ttype: string): WidgetOption[] {
  return FALLBACK[ttype] ?? [];
}

export type FieldModifierMode = "off" | "always" | "domain";

export function modifierToMode(value: boolean | string | undefined): FieldModifierMode {
  if (value === true || value === "1") return "always";
  if (typeof value === "string" && value.trim() && value !== "0") return "domain";
  return "off";
}

export function modeToModifier(mode: FieldModifierMode, domain: string): boolean | string | undefined {
  if (mode === "always") return true;
  if (mode === "domain") {
    const d = domain.trim();
    return d && d !== "[]" ? d : undefined;
  }
  return undefined;
}
